File: packages/typescript/src/streaming.ts
import type { Event, ProviderId, Session } from "./types.js";

const TERMINAL_STATUSES = new Set(["completed", "failed", "cancelled"]);

export async function* pollingStream(options: {
  provider: ProviderId;
  sessionId: string;
  fetchSession: () => Promise<Session>;
  fetchNewEvents: (cursor: string | undefined) => Promise<[Event[], string | undefined]>;
  pollBase?: number;
  pollCap?: number;
  keepaliveSecs?: number;
}): AsyncIterableIterator<Event> {
  const {
    provider,
    sessionId,
    fetchSession,
    fetchNewEvents,
    pollBase = 1000,
    pollCap = 15000,
    keepaliveSecs = 30,
  } = options;

  let seq = 0;
  let idleCount = 0;
  let cursor: string | undefined;
  let lastActivity = Date.now();

  while (true) {
    const session = await fetchSession();
    const [events, nextCursor] = await fetchNewEvents(cursor);
    if (nextCursor !== undefined) cursor = nextCursor;

    for (const ev of events) {
      ev.seq = seq++;
      idleCount = 0;
      lastActivity = Date.now();
      yield ev;
    }

    if (TERMINAL_STATUSES.has(session.status)) {
      yield {
        kind: "status",
        sessionId,
        provider,
        ts: new Date().toISOString(),
        seq: seq++,
        data: { value: session.status },
      };
      return;
    }

    const now = Date.now();
    if ((now - lastActivity) / 1000 > keepaliveSecs) {
      yield {
        kind: "provider_event",
        sessionId,
        provider,
        ts: new Date().toISOString(),
        seq: seq++,
        data: { type: "keep_alive" },
      };
      lastActivity = now;
    }

    await sleep(Math.min(pollCap, pollBase * Math.pow(2, Math.min(idleCount, 4))));
    idleCount = events.length > 0 ? 0 : idleCount + 1;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function* claudeSseStream(
  baseUrl: string,
  apiKey: string,
  extraHeaders: Record<string, string>,
  sessionId: string,
  sinceProcessedAt: string | undefined,
): AsyncIterableIterator<Record<string, unknown>> {
  const seenIds = new Set<string>();
  const headers: Record<string, string> = {
    "x-api-key": apiKey,
    ...extraHeaders,
  };

  if (sinceProcessedAt) {
    const url = new URL(`${baseUrl}/v1/sessions/${sessionId}/events`);
    url.searchParams.set("processed_at_gt", sinceProcessedAt);
    const resp = await fetch(url.toString(), { headers });
    if (resp.ok) {
      const data = (await resp.json()) as { events?: Array<Record<string, unknown>> };
      for (const ev of data.events ?? []) {
        const id = ev["id"] as string | undefined;
        if (id && !seenIds.has(id)) {
          seenIds.add(id);
          yield ev;
        }
      }
    }
  }

  const streamUrl = `${baseUrl}/v1/sessions/${sessionId}/events/stream`;
  const resp = await fetch(streamUrl, { headers });
  if (!resp.ok || !resp.body) return;

  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";
    for (const line of lines) {
      if (!line.startsWith("data: ")) continue;
      const payload = line.slice(6).trim();
      if (payload === "[DONE]") return;
      try {
        const data = JSON.parse(payload) as Record<string, unknown>;
        const id = data["id"] as string | undefined;
        if (id && seenIds.has(id)) continue;
        if (id) seenIds.add(id);
        yield data;
      } catch {
        // malformed SSE frame — skip
      }
    }
  }
}

File: packages/python/src/omniharness/_streaming.py
from __future__ import annotations

import asyncio
import json
from datetime import datetime, timezone
from typing import AsyncIterator, Awaitable, Callable, Optional

import httpx

from .types import Event, ProviderId, Session

TERMINAL_STATUSES = {"completed", "failed", "cancelled"}


async def polling_stream(
    *,
    provider: ProviderId,
    session_id: str,
    fetch_session: Callable[[], Awaitable[Session]],
    fetch_new_events: Callable[[Optional[str]], Awaitable[tuple[list[Event], Optional[str]]]],
    poll_base: float = 1.0,
    poll_cap: float = 15.0,
    keepalive_secs: float = 30.0,
) -> AsyncIterator[Event]:
    seq = 0
    idle_count = 0
    cursor: Optional[str] = None
    last_activity = asyncio.get_event_loop().time()

    while True:
        session = await fetch_session()
        events, cursor = await fetch_new_events(cursor)

        for ev in events:
            ev.seq = seq
            seq += 1
            idle_count = 0
            last_activity = asyncio.get_event_loop().time()
            yield ev

        if session.status in TERMINAL_STATUSES:
            yield Event(
                kind="status",
                session_id=session_id,
                provider=provider,
                ts=datetime.now(timezone.utc),
                seq=seq,
                data={"value": session.status},
            )
            return

        now = asyncio.get_event_loop().time()
        if now - last_activity > keepalive_secs:
            yield Event(
                kind="provider_event",
                session_id=session_id,
                provider=provider,
                ts=datetime.now(timezone.utc),
                seq=seq,
                data={"type": "keep_alive"},
            )
            seq += 1
            last_activity = now

        await asyncio.sleep(min(poll_cap, poll_base * (2 ** min(idle_count, 4))))
        idle_count = 0 if events else idle_count + 1


async def claude_sse_stream(
    http: httpx.AsyncClient,
    session_id: str,
    since_processed_at: Optional[str],
) -> AsyncIterator[dict]:  # type: ignore[type-arg]
    seen_ids: set[str] = set()

    if since_processed_at:
        resp = await http.get(
            f"/v1/sessions/{session_id}/events",
            params={"processed_at_gt": since_processed_at},
        )
        for ev in resp.json().get("events", []):
            if ev["id"] not in seen_ids:
                seen_ids.add(ev["id"])
                yield ev

    async with http.stream("GET", f"/v1/sessions/{session_id}/events/stream") as resp:
        async for line in resp.aiter_lines():
            if not line.startswith("data: "):
                continue
            payload = line[6:]
            if payload.strip() == "[DONE]":
                break
            data = json.loads(payload)
            ev_id = data.get("id", "")
            if ev_id and ev_id in seen_ids:
                continue
            if ev_id:
                seen_ids.add(ev_id)
            yield data
